src/rust/bitbox-aes/src/lib.rs

```diff
@@ -0,0 +1,3065 @@
+// Copyright 2025 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![no_std]
+
+#[cfg_attr(test, macro_use)]
+extern crate alloc;
+
+use alloc::vec::Vec;
+
+use hmac::{digest::FixedOutput, Mac, SimpleHmac};
+use sha2::Sha256;
+
+// AES block size.
+const BLOCK_SIZE: usize = 16;
+
+fn pad_length(buf_len: usize) -> usize {
+    BLOCK_SIZE - buf_len % BLOCK_SIZE
+}
+
+fn pkcs7_pad(buf: &mut Vec<u8>) {
+    let pad_len = pad_length(buf.len());
+    buf.resize(buf.len() + pad_len, pad_len as u8);
+}
+
+fn pkcs7_unpad(buf: &mut Vec<u8>) -> Result<(), ()> {
+    if buf.len() % BLOCK_SIZE != 0 {
+        return Err(());
+    }
+    let pad_len = *buf.last().ok_or(())? as usize;
+    if pad_len == 0 || pad_len > BLOCK_SIZE {
```

Comment:
Isn't this a programmer error? i.e. shouldn't you panic? (can also use `debug_assert` or `assert` of course).

Reply:
Also the caller can always `.unwrap()` if the caller prefers a panic over graceful error handling. The caller should be able to decide that. 

It's sth I dislike about the rust std: half the functions return Result/Option, the other half panics, it's quite inconsistent. But Result/Option is always better, the caller can always unwrap if they are sure their input is good.

Reply:
Result/Option is always less performant and generates more code. I don't think the library should force the developer to use `unwrap`. I think code bases in general should be `unwrap`-free unless they are "work in progress".

stdlib will panic when it makes sense, like if allocation fails or if indexing fails.

Reply:
So errors that are due to bugs are fine to panic on.

Reply:
In this specific case I guess it isn't a programmer bug. But if you don't return the cause for the error then the library user will have to write all the logic for figuring out what is wrong to inform the end-user?

Reply:
I don't disagree in principle, a concrete error enum with different failure cases makes sense if the caller needs to now. But in our case the caller does not care, and within our project I don't care much for adding types just for the sake of it if every call-site just discards it. 

At least in the original C implementation the callsites don't care about this error - if you feel like the errors should be handled specifically based on the failure case, it's a separate discussion.

Reply:
> So errors that are due to bugs are fine to panic on.

I could agree, but even with programmer bugs, you'd prefer an API call to fail gracefully in case of a bug instead of crashing everything.

But "programmer bug" is not clearly defined. Is it a programmer bug if the flash memory contains something unexpected? Is it the caller's job to verify the input it's valid or is it this module's job? I don't know if there is a clear line.